br/pkg/lightning/restore/data_sample_check.go

```diff
@@ -0,0 +1,307 @@
+		zap.String("file", fileMeta.Path),
+		zap.Int64("offset", offset),
+		zap.String("error", errMsg))
+}
+
+func (d *dataSampleCheck) checkRoutine(ctx context.Context, fileChan chan *sampledDataFileInfo, errChan chan error) {
+	defer d.wg.Done()
+
+	var resultErr error
+	defer func() {
+		errChan <- errors.Trace(resultErr)
+	}()
+
+	rc := d.controller
+	for fileInfo := range fileChan {
+		fileMeta := fileInfo.DataFile.FileMeta
+		fileParser, err := rc.createDataFileParser(ctx, fileMeta, config.DefaultCSVDataInvalidCharReplace)
+		if err != nil {
+			if errors.Cause(err) != context.Canceled {
+				resultErr = err
+			}
+			return
+		}
+
+		ignoreColumns, err := rc.cfg.Mydumper.IgnoreColumns.GetIgnoreColumns(fileInfo.Table.DB, fileInfo.Table.Name, rc.cfg.Mydumper.CaseSensitive)
+		if err != nil {
+			resultErr = err
+			return
+		}
+		ignoreColumnsMap := ignoreColumns.ColumnsMap()
+
+		ddlColumnMap := make(map[string]*model.ColumnInfo, len(fileInfo.TableInfo.Columns))
+		for _, col := range fileInfo.TableInfo.Columns {
+			ddlColumnMap[col.Name.L] = col
+		}
+		var columnNames []string
+		var rowsChecked, columnCountMismatchRows, invalidCharRows, lastOffset int64
+		columnCount := -1
+		var count int64
+		for {
+			count++
+			if d.checkCfg.Rows != config.CheckAllRows && count > d.checkCfg.Rows {
+				break
+			}
+			err = fileParser.ReadRow()
+			if err != nil {
+				if errors.Cause(err) == io.EOF {
+					break
+				}
+
+				if errors.Cause(err) != context.Canceled {
+					resultErr = err
+				}
+				fileParser.Close()
+				return
+			}
+
+			row := fileParser.LastRow()
+			if columnCount == -1 {
+				columnNames = fileParser.Columns()
+				if len(columnNames) == 0 {
+					columnCount = 0
+					columnNames = make([]string, 0, columnCount)
+					for _, col := range fileInfo.TableInfo.Columns {
+						if !col.Hidden {
+							columnNames = append(columnNames, col.Name.L)
+							columnCount++
+						}
+					}
+				} else {
+					columnCount = len(columnNames)
+				}
+			}
+			rowsChecked++
+			if len(row.Row) != columnCount {
+				columnCountMismatchRows++
+				d.logCheckFailedRow(row.Row, fileMeta, lastOffset,
+					fmt.Sprintf("column count mismatch, expect %d, got %d", columnCount, len(row.Row)))
+			}
+
+			for colIdx, col := range row.Row {
+				if colIdx >= columnCount {
+					break
+				}
+
+				// no need to check ignored column, since it's not encoded in kv
+				colName := columnNames[colIdx]
+				if _, ok := ignoreColumnsMap[colName]; ok {
+					continue
+				}
+				columnInfo := ddlColumnMap[colName]
+				if columnInfo == nil {
+					// we don't check this kind of error here
+					continue
+				}
+				colType := columnInfo.FieldType
+				// we only check non-binary string types here
+				if !types.IsNonBinaryStr(&colType) {
+					continue
+				}
+				colStr := col.GetString()
+				if !utf8.ValidString(colStr) || strings.Contains(colStr, config.DefaultCSVDataInvalidCharReplace) {
```

Comment:
Since `DefaultCSVDataInvalidCharReplace` itself is a valid utf8 charcter, what if the origin input file itself contains this char, e.g, the input file is from an upstream that it does the same replacement for legacy invalid data.

Reply:
in most case, this should not happen. if it's the case, this check can let user know and check themself.